common/unsafe/src/main/java/org/apache/spark/unsafe/array/ByteArrayMethods.java

```diff
@@ -40,24 +41,41 @@ public static int roundNumberOfBytesToNearestWord(int numBytes) {
     }
   }
 
+  private static final boolean alignedArch = UnsafeAlignedOffset.getAlignedArch();
   /**
    * Optimized byte array equality check for byte arrays.
    * @return true if the arrays are equal, false otherwise
    */
   public static boolean arrayEquals(
       Object leftBase, long leftOffset, Object rightBase, long rightOffset, final long length) {
     int i = 0;
-    while (i <= length - 8) {
-      if (Platform.getLong(leftBase, leftOffset + i) !=
-        Platform.getLong(rightBase, rightOffset + i)) {
-        return false;
+
+    // check if stars align and we can get both offsets to be aligned
+    if ((leftOffset % 8 != 0) && ((leftOffset % 8) == (rightOffset % 8))) {
+      while (leftOffset % 8 != 0 && i < length) {
+        if (Platform.getByte(leftBase, leftOffset + i) !=
+            Platform.getByte(rightBase, rightOffset + i)) {
+              return false;
+        }
+        i += 1;
+      }
+    }   
+    // for architectures that suport unaligned accesses, chew it up 8 bytes at a time
+    if (!alignedArch || ((leftOffset % 8 == 0) && (rightOffset % 8 == 0))) {
```

Comment:
This is slick, yes. It lets aligned arch take advantage of the faster compare where possible for just a little extra check. It's also no extra work x64 et al, because it's not going to cause _more_ byte-by-byte comparison. There always 0-7 of them. But it does let the reads become aligned which is still a bit faster on x86.